### ISSUE TITLE
Investigate use/recommendation of 'provided' scope for connectors

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#4455]
tbd

**Comment by planetf1** [comment]
Currently we only use 'provided' scope for gaian & derby - since both relate to libraries we do not ourselves ship, and instead require them to be in-situ on the target (somewhat of a similar scenario)

Gradle lends itself better to a broader range of configurations -- for example we could far more easily lay out seperate trees in the assembly for 'clients', server, utilities, and in each case define the core common set that would be suited as 'provided'.

Draft set of tasks
 - [ ] Identify core set for server chassis
 - [ ] Do we need a maven bom to reflect
 - [ ] checking if any build ordering impact
 - [ ] ensuring assembly has all 'provided' libraries ie in lib (ideally making use of scope)
 - [ ] updating build of connectors accordingly
 - [ ] reflecting same in gradle
 - [ ] looking at commonality for clients

**Comment by davidradl** [comment]
Makes sense. One comment. I assume that when we have multiple database connectors, there could / would be clashing classes e.g. different jdbc drivers. And different connectors would depend on different parts of the Server Chassis. For integration connectors and governance action framework connectors. I am thinking we should group the 'provided' library content per consuming connector, for ease of consumption.

**Comment by planetf1** [comment]
@davidradl I think the jdbc drivers themselves will have unique names. if they have an uber jar with other dependencies there is a risk (there always is) of divergent versions, though those libs may use shading - worthy of more checks. But I don't think the top level driver per se will be an issue.

For the server chassis, even though connectors may depend on different egeria modules at runtime, I'm thinking for simplicity we need to define a limited set of 'platforms' - maybe only one - that will ALWAYS be there.

However Most recently we have created a stripped down chassis which only has the very minimal capabilities. This can be assumed to always be there, but many other egeria modules something like a view service may depend on may, or may not, be there. 

In that case whilst we can use 'provided' scope in the build process, we need more... We need to ensure code is always dynamically loaded, and that errors are sensibly detected and reported, as deciding what is present in the chassis at runtime becomes more of a deployment question

In essence we need to move the focus on that assembly from the build process to deployment - but to do this in such a way that anyone deploying can do it fairly easy. With say 5, 10 modules that might be viable. with 200-400+ (in egeria) we need more.

Thinking about this more, provided scope is only part of the approach - it adds the build flexibility, but doesn't solve the problem of the deployer getting the environment right.

**Comment by bogdan-sava** [comment]
No need to have dependencies at all.
At runtime, if jar is missing from the lib folder,  will ClassNotFoundException will be catch and wrapped and logged in audit log.

For that I need to change a little the https://github.com/odpi/egeria/pull/4347 . In the PR I moved the dependencies in profile, but if they are removed works just fine if the JARs are in the lib folder. 
Already test that.

PS: I kept dependencies in the PR to be able to merge it before other changes.

**Comment by davidradl** [comment]
@bogdan-sava I am looking of `https://github.com/odpi/egeria-connector-ibm-information-server/blob/master/datastage-adapter/src/main/java/org/odpi/egeria/connectors/ibm/datastage/dataengineconnector/DataStageConnector.java` which has an` import org.odpi.openmetadata.frameworks.connectors.properties.ConnectionProperties;`

So this is a dependancy that needs to be resolved at compile time. It is brought in using 
```
   <dependency>
                <groupId>org.odpi.egeria</groupId>
                <artifactId>open-connector-framework</artifactId>
                <version>${open-metadata.version}</version>
                <scope>provided</scope>
            </dependency>
```

This means that the dependancy should be provided at runtime. 

So without the provided dependancy in the pom file it would not compile.

**Comment by planetf1** [comment]
Yes - provided is needed for compile time (there will be a gradle equiv). 


The main assembly I think also should be changed so that we provide the dependencies in server/lib or similar - not just connectors, but all. The chassis should be much leaner by default - as indeed you added in optional profile definition @bogdan-sava . I don't think we need the uber jar for the chassis long term, though do need to be ok with the dev use case also.

Note though some experimentation/investigation is needed here -- for example if we remove dependencies, we also - in maven at least - remove the control of build ordering. We also change how the assembly is built, as it may no longer know what dependencies to include, so need to adopt an alternative mechanism for this (gradle may make this easier)

I think we should start considering making this change in the next, or next+1 dev iteration ie Feb/Mar.

**Comment by planetf1** [comment]
When looking at the postgres connector specifically & trying a fix for this generic issue there, I note that gradle now has 'compileOnly' which is pretty much what we want for 'provided' :-)

However clarification over what can be assumed to be present, modifying assemblies to include all the libs as jars (not an uber jar), plus docs still need to be worked here, ie in the base

**Comment by cmgrote** [comment]
I would advocate for the approach of using the `provided` scope and still allowing a semi-uber-jar for the connector to be built (with its unique dependencies, but leaving out the Egeria dependencies), for these reasons:

- This will ensure that we have a singular "release" with pre-defined depedencies (critically versions) baked-in
- Should significantly improve the ability for us to identify issues and debug them accordingly (rather than needing to second-guess what versions of libraries someone might be running alongside the connector)
- Will also mean that our code analysis mechanisms (security, etc) will identify potential issues with the baked-in dependencies (don't think there'd be any way for us to do this if users are manually choosing their dependency versions in the `/lib` directory?)

This is the approach I've taken in both the IGC and Atlas connectors (at least, explicitly including only the non-Egeria dependencies in the uber jar they each build, though I may not be explicitly using the `provided` scope)...

**Comment by planetf1** [comment]
Also note that any recommendations should comment on use of snapshots (and usr of -U or the gradle equiv) - see https://github.com/odpi/egeria-database-connectors/issues/14 . This is more relevant in the initial stages of building a new connector alongside core dev work, whilst when using a stable base using the last released version of egeria is 100% the way to go.

Leaving any further eval/doc on this until we've made base packaging changes ~ Feb 2021 to unpack dependencies in our assembly

**Comment by planetf1** [comment]
This is part of the work to slim down the chassis - @bogdan-sava assigned to you to manage/coordinate but feel free to reassign particular changes needed back once you've figured out the overall approach.

**Comment by planetf1** [comment]
I think what we need to make this work (an what is tried to be suggested above) is

 * Clarity on what the base platform is -- ie what can we always assume is present. This can be a packaging pom (we'd need to exclude this from the dependency check)
 * Use of 'provided' or 'compileOnly' for the scope of this 'base platform' dependency
 * Building jar-with-dependencies OR inclusion in 'lib' directory of all additional modules that are regular dependencies (ie compile or Implementation) (this is where the extra dependencies needed for a particular connector come in)
 
 This would extend not just to connectors, but everything outside the core platform, so for example 'optional' OMAS modules etc.
 
 I think we're both saying the same thing?

**Comment by planetf1** [comment]
Other potential impacts
 * Maven assembly
 * Gradle assembly
 * CI/CD process - publishing of artifacts
 * Docs around use of additional connectors
 * Developer information on compile/debug ie in intellij (but also cli)
 
 If this direction sounds good I"m fine we close this issue, and create a new clean issue to track the implementation. This can be a high level issue to which we attach child issues (those can then be independently assigned & worked on & merged). First step being to edit a sequence of steps to get us to the goal. 
 
 One thing I'm wondering is whether to skip the maven version of this and just go gradle. the downside is it could delay the work as gradle builds needs verification/some completion. However last time we rejigged this area the maven assembly/build was problematic. I'm somewhat neutral.
